src/main/java/net/logstash/logback/util/ReusableJsonFormatterPool.java

```diff
@@ -0,0 +1,223 @@
+ * collected by the JVM when running low in memory.
+ * 
+ * @author brenuart
+ */
+public class ReusableJsonFormatterPool<Event extends DeferredProcessingAware> {
+
+    /**
+     * The minimum size of the byte buffer used when encoding events.
+     */
+    private final int minBufferSize;
+    
+    /**
+     * The factory used to create {@link JsonFormatter} instances
+     */
+    private final CompositeJsonFormatter<Event> formatterFactory;
+    
+    /**
+     * The pool of reusable JsonFormatter instances.
+     * May be cleared by the GC when running low in memory.
+     * 
+     * Note:
+     *   JsonFormatters are not explicitly disposed when the GC clears the SoftReference. This means that
+     *   the underlying Jackson JsonGenerator is not explicitly closed and the associated memory buffers
+     *   are not returned to Jackson's internal memory pools.
+     *   This behavior is desired and makes the associated memory immediately reclaimable - which is what
+     *   we need since we are "running low in memory".
+     */
+    private volatile SoftReference<Deque<ReusableJsonFormatter>> formatters = new SoftReference<>(null);
+    
+
+    public ReusableJsonFormatterPool(CompositeJsonFormatter<Event> formatterFactory, int minBufferSize) {
+        this.formatterFactory = Objects.requireNonNull(formatterFactory);
+        this.minBufferSize = minBufferSize;
+    }
+
+    /**
+     * A reusable JsonFormatter holding a JsonFormatter writing inside a dedicated {@link ReusableByteBuffer}.
+     * Closing the instance returns it to the pool and makes it available for subsequent usage, unless the
+     * underlying {@link CompositeJsonFormatter.JsonFormatter} threw an exception during its use.
+     * 
+     * <p>Note: usage is not thread-safe.
+     */
+    public class ReusableJsonFormatter implements Closeable {
+        private final ReusableByteBuffer buffer;
+        private final CompositeJsonFormatter<Event>.JsonFormatter formatter;
+        private boolean recyclable = true;
+        
+        ReusableJsonFormatter(ReusableByteBuffer buffer, CompositeJsonFormatter<Event>.JsonFormatter formatter) {
+            this.buffer = buffer;
+            this.formatter = formatter;
+        }
+
+        /**
+         * Return the underlying buffer into which the JsonFormatter is writing.
+         * 
+         * @return the underlying byte buffer
+         */
+        public ReusableByteBuffer getBuffer() {
+            return buffer;
+        }
+        
+        /**
+         * Write the Event in JSON format into the enclosed buffer using the enclosed JsonFormatter.
+         * 
+         * @param event the event to write
+         * @throws IOException thrown when the JsonFormatter has problem to convert the Event into JSON format
+         */
+        public void write(Event event) throws IOException {
+            try {
+                this.formatter.writeEvent(event);
```

Comment:
What is the behavior if write is called after this ReusableJsonFormatter has been disposed of, or marked as not recyclable?  Should write throw an exception if this ReusableJsonFormatter has been disposed or if it has been marked as not recyclable before writing?

Reply:
> behaviour after it has been disposed

See https://github.com/logstash/logstash-logback-encoder/pull/631#discussion_r704724530

> behaviour when marked as not recyclable

JsonGenerator are stateful in the sense it knows for instance if the current structure of what has been generated so far allows for an object to be serialised. It would for instance complain if you attempt to `writeObject()` if a field name has not been output previously. There is unfortunately no way to "reset" it and revert to a clean initial state. The only solution I found is to avoid reusing it after it threw an exception. The "recyclable" field captures this information for when the ReusableJsonFormatter is released.
It can theoretically also throw an IOException if it has problem to write in the underlying OutputStream. This will never happen in our case.

Now, regarding your question... the current usage pattern is such that a ReusableJsonFormatter is used to write only one Event before it is released. But nothing prevents you from writing two events in a row with the same instance, they will be both written in the same buffer (that you can reset between the two write attempts). If you do so, I must say the outcome of the second write operation is unpredictable after the first threw an exception... It all depends on the JsonGenerator and what you are about to write.

I decided to keep the same behaviour has the JsonGenerator itself. The net result is similar to the following example when the first writeObjectField() operation throws an IOException:

```
JsonGenerator gen = ...
try {
  gen.writeObjectField("field1", obj1);
} catch(IOException e) {
  // ignore
}

gen.writeObjectField("field2", obj2);
```
... what would be the result? I don't know.. I haven't tested it... but such code doesn't look very wise :-(

Anyway, I can make the implementation more robust and, for instance, deny any further write() operation after an exception is thrown...

Reply:
Ok.  Thanks for the explanation. :)

I think it's fine as-is.  If this were a user-facing class, I'd say we'd need to add some more defensive logic and better error messages.  But since this is internal to the library, we can keep it simple.